internal/spdxlicense/generate/generate_license_list.go

```diff
@@ -102,24 +126,36 @@ func run() error {
 	return nil
 }
 
-// Parsing the provided SPDX license list necessitates a two pass approach.
-// The first pass is only related to what SPDX considers the truth. These K:V pairs will never be overwritten.
+// Parsing the provided SPDX license list necessitates a three pass approach.
+// The first pass is only related to what SPDX considers the truth. We use license info to
+// find replacements for deprecated licenses.
 // The second pass attempts to generate known short/long version listings for each key.
 // For info on some short name conventions see this document:
 // https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/#license-short-name.
 // The short long listing generation attempts to build all license permutations for a given key.
 // The new keys are then also associated with their relative SPDX value. If a key has already been entered
 // we know to ignore it since it came from the first pass which is considered the SPDX source of truth.
 // We also sort the licenses for the second pass so that cases like `GPL-1` associate to `GPL-1.0` and not `GPL-1.1`.
+// The third pass is for overwriting deprecated licenses with replacements, for example GPL-2.0+ is deprecated
+// and now maps to GPL-2.0-or-later.
 func processSPDXLicense(result LicenseList) map[string]string {
 	// first pass build map
 	var licenseIDs = make(map[string]string)
+	var overwrite = make(map[string]string)
+
 	for _, l := range result.Licenses {
 		cleanID := strings.ToLower(l.ID)
 		if _, exists := licenseIDs[cleanID]; exists {
 			log.Fatalf("duplicate license ID found: %q", cleanID)
 		}
 		licenseIDs[cleanID] = l.ID
+
+		if l.Deprecated {
+			replacement := findEquivalentLicense(l, result.Licenses)
+			if replacement != nil {
+				overwrite[cleanID] = replacement.ID
+			}
+		}
```

Comment:
Is there a reason we're building this `overwrite` collection here and waiting to use it until later? What's the reason for not replacing deprecated licenses right away before we generate more permutations below?

Reply:
Good question: The order of variations/permutations of a license ID matters because of we how shuffle its digits, that is because the permutation code can generate the same value for two difference licenses, for example: The licenses `ABC-1.0` and `ABC-1.1` can both map to `ABC-1`, so we need to guarantee the order they are created to avoid mapping them wrongly. So we use a sorted list.

To overwrite deprecated licenses during the first pass we would later on rely on map order, [which in go is not consistent by design](https://stackoverflow.com/a/55925880).

I think we should revisit how we permute license IDs, but that is for another PR, as this code generates our global list of license IDs and that is delicate.